Proposed fix for #8499

diego-ninja opened the pull request:
Changed lock function to compare timestamps instead of DateTimeInterface objects directly.

When using optimistic lock with DateTimeInterface based version field a bug appears due to the use of the === operator for comparing the lock version and the entity version. This comparison always resolves to false because the === operator when comparing objects is only true when both sides are the exact same instance of the object.

To fix the issue I have decided to compare timestamps instead the DateTimeInterface based objects directly, calling getTimestamp() method and doing a strict comparison.

Modified OptimisticLockException to use DateTimeInterface instead of DateTime class.

Added test suite to cover case.

Fixes #8499

greg0ire commented:
Please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#working-on-topic-branches).

diego-ninja commented:
> Please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#working-on-topic-branches).

Done.

greg0ire commented:
I meant the _commit_ message, so you can squash your current commits together, and reuse your PR message as a commit message. BTW, when you create a PR with just one commit, Github will reuse its message as a PR message.

diego-ninja commented:
> I meant the _commit_ message, so you can squash your current commits together, and reuse your PR message as a commit message. BTW, when you create a PR with just one commit, Github will reuse its message as a PR message.

I hope it's ok now.

greg0ire commented:
I will do it for you.

greg0ire commented:
:thinking: is this line wrong: https://github.com/doctrine/orm/blob/e40ac3e1d0f82e39c8ae9990866773372a4a069d/lib/Doctrine/ORM/EntityManagerInterface.php#L218 ?

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH8499Test.php:
```diff
@@ -0,0 +1,175 @@
+        $this->_em->persist($entity);
+        $this->_em->flush();
+        $this->_em->refresh($entity);
+        $lastVersion = $entity->getRevision();
+
+        $this->assertNotEquals($firstVersion, $lastVersion, 'Version field value not updated on persist');
```
Let's be consistent with the test below?
```suggestion
        $this->assertNotEquals($firstVersion, $entity->getRevision(), 'Version field value not updated on persist');
```

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH8499Test.php:
```diff
@@ -0,0 +1,175 @@
+        $this->_em->flush();
+        $this->_em->refresh($entity);
+        $firstVersion = $entity->getRevision();
+        sleep(1);
```
I wish there was a better way to do this, maybe another maintainer will have a better idea. Right now, I think this test is probably the longest test of the whole test suite.

greg0ire replied:
I have given it more thought:
That feature also works with integer Versions.
You already tested that the version was initialized on insert. If an entity's integer version is updated on update, I think we can reasonably think it will work too for a datetime version.

1. Are we sure such a test does not already exist?
2. Should we rewrite that particular test with an integer field so that we can avoid that call to `sleep`? And if yes, maybe it shouldn't be in `GH9499Test`. This test passes without your fix, doesn't it?

beberlei replied:
Use dbal to update the row into the past

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH8499Test.php:
```diff
@@ -0,0 +1,173 @@
+        $this->_em->persist($entity);
+        $this->_em->flush();
+
+        $this->assertNotEquals($firstVersion, $entity->getRevision(), 'Version field value not updated on persist');
```
We already tested that above, right? IMO, we should test the locking mechanism independently from version update. Could we achieve that by adding a setter for the version field and calling it at the right time?

diego-ninja replied:
Yeah, your right, stritctly speaking I should only test two things, that the lock function works with a correct dataset and throws a exception when not. So the rest of the test cases are not needed here, I mean, the case of the version field being initailized and updated in a proper way should be tested anywhere else but not here. So, I'm going to remove the "extra" cases from the test suite.

sounds good to you?

diego-ninja replied:
Okie, I'm checking the OptimisticTest.php included in Tests\ORM\Functional\Locking directory again, there are a bunch of tests which covers all cases except the one with lock mechanism working as expected, the one that would have detected this bug, So I'm going to write that test only. 

it's that okie?

greg0ire replied:
It's super okie 👍 Thanks a lot for investigating this.

diego-ninja replied:
It's done, I write the test without using sleep function, following the work done in OptimisticTest.php. Check it, I think is a good approach, I followed the @beberlei advice and modified the version datetime in the past using dbal.

diego-ninja committed:
Changed lock function to compare timestamps instead of DateTimeInterface objects directly.

When using optimistic lock with DateTimeInterface based version field a bug appears due to the use of the === operator for comparing the lock version and the entity version. This comparison always resolves to false because the === operator when comparing objects is only true when both sides are the exact same instance of the object.

To fix the issue I have decided to compare timestamps instead the DateTimeInterface based objects directly, calling getTimestamp() method and doing a strict comparison.

Modified OptimisticLockException to use DateTimeInterface instead of DateTime class.

Added test suite to cover case.

Fixes #8499

greg0ire reviewed (approved):
Looks great 👍
